unison-core/src/Unison/Name.hs

```diff
@@ -178,9 +199,73 @@ fromSegment = unsafeFromText . NameSegment.toText
 segments :: Name -> [NameSegment]
 segments (Name n) = NameSegment <$> segments' n
 
+reverseSegments :: Name -> [NameSegment]
+reverseSegments (Name n) = NameSegment <$> NameSegment.reverseSegments' n
+
 countSegments :: Name -> Int
 countSegments n = length (segments n)
 
+-- The `Ord` instance for `Name` considers the segments of the name
```

Comment:
Should we use a newtype around Name for this Ord logic that’s kind of surprising?  Like, we have an application that’s more efficient with this new Ord, but wouldn’t we quickly find a second application that this Ord is not good for?

Reply:
Yeah, I considered that, but then, I think that the purpose of an `Ord` instance is to be computationally useful. The current byte ordering isn't useful for any purpose we have right now - not for indexing, and not for display-level alphabetical sorting (byte ordering is wrong for that as well). So I don't think there's a good argument for it being the default. The current ordering is just the first ordering we happened to think of given the representation as a newtype around `Text`. :)

IMO we have been too influenced by that rather arbitrary representation choice - if we have thought of suffix-based resolution and search as being important things for names, we might have represented it as a snoc list of segments, and then that representation's "natural" ordering would actually coincide with what this PR is doing. I think it could be cool to try swapping out the representation (in a separate PR), but I didn't try for that here.

That said, we can definitely revisit what the default is if/when we have other competing orderings for `Name` that are useful! :)

Reply:
Maybe another POV here is that what an `Ord` instance buys you over just a hashing function is the ability to do meaningful range queries, and the current `Ord` isn't letting us do any range query we care about - it lets us do prefix matching based on the byte order of the name, which is not very useful.